### ISSUE TITLE
Add sms support for `aws_sns_topic_subscription`

### ISSUE THREAD

**Issue by hugorut** [infracost/infracost#1393]
[](https://aws.amazon.com/sns/sms-pricing/)Currently infracost only supports `http|https` message delivery for `aws_sns_topic_subscription`. Lots of infracost users use sns to deliver sms messages to their end users. Infracost should figure out a way to support this use case.

## Pricing

AWS SMS pricing information can be found here: https://aws.amazon.com/sns/sms-pricing/ .  For the `aws_sns_topic_subscription` use case, Infracost would only be interested in outbound costs.  

For outbound sms prices varies by destination country. At face value this seems easy to do, however there are a few things that make implementing SMS costs problematic.

### Per country billing variance

The price and components for sending SMS messages varies between countries. For example see below:

**US**
  
<img width="882" alt="Screenshot 2022-02-23 at 11 53 54" src="https://user-images.githubusercontent.com/6455139/155326843-b8d4af19-c9a9-4ec4-9f61-bc7775398f7e.png">

**Spain**

<img width="881" alt="Screenshot 2022-02-23 at 12 00 40" src="https://user-images.githubusercontent.com/6455139/155326794-247474e4-4c85-46b0-aa8d-742ede213879.png">

**India**

<img width="873" alt="Screenshot 2022-02-23 at 11 54 01" src="https://user-images.githubusercontent.com/6455139/155326808-955e3dba-9712-4af0-8dbf-e0858ff3333c.png">

AWS also states that pricing can also vary between carrier in country (although from our investigation none of the inter country carriers have a different price). 

The lack of consistency in pricing means that offering Infracost users all encompassing configuration options for sms pricing could be quite complex for a user to configure. Below I outline a few examples.

### Complex usage parameters

We could modify the `monthly_requests` usage param to be a more complex array type which would allow users to specify sms destination info. Take the following examples.

```bash
aws_sns_topic.my_sns_topic:
  monthly_requests: 
    - destination: "us" # The destination of the sms requests, must be a valid region/country.
      sent: 1000000 # Monthly requests which are sent to US based numbers.
      sms_identity: "10dlc" # SMS Origination Identities, valid options are 10dlc, longcode, shortcode, tollfree.
      sms_type: "transactional" # The type of sms sent, can be one of transactional or promotional
      # carrier: "some carrier" can be used to specify the destination carrier	
```

This approach would need us to modify the pricing api to incorporate custom scrapers for outbound sms pricing, found here: [https://s3.amazonaws.com/aws-messaging-pricing-information/TextMessageOutbound/prices.json](https://s3.amazonaws.com/aws-messaging-pricing-information/TextMessageOutbound/prices.json) . 

**Pros**

- The most flexible way to specify outbound sms costs
- Automatically adjusts to aws pricing changes

**Cons**

- Requires complex validation for destination rules
- Difficult for users to understand

### Static usage cost

Another approach would be to let users to specify a static cost for the `monthly_request` sns usage param. This would enable users to provide their own estimate for outbound sms costs. Usage could look like such:

```bash
aws_sns_topic.my_sns_topic:
    monthly_requests: 1000
    sms_req_cost: 0.0058
```

**Pros**

- Simple to implement
- Easy to understand

**Cons**

- Requires users to keep up to date with monthly pricing
- Does not change per country/region so users need to provide an average - therefore a worse estimate of costs incurred

---

A third approach could lie somewhere between both approaches. This could mean offering more complex usage options but removing option to specify additional usage params like `sms_type` or `carrier`. In these cases Infracost could assume a set scenario, however we would need to communicate this with the users. For example 

*infracost only supports transactional sms sending costs* 

or 

*infracost  does not support inter carrier sending, we assume a single cost for a country*

**Comment by tim775** [comment]
@hugorut @aliscott 
Great write up Hugo!  After looking into this a bit here's my opinion on what we should do:
1.  Go with the "Static usage cost" approach outlined above.  Use a default price for SMS notifications while letting users override that price with a usage param if needed.  The CPAPI actually has a price "$0.75 per 100 Amazon SNS SMS Notifications thereafter" which we can use as the default.
2. `aws_sns_topic_subscription` isn't the right resource to track the costs, we should put them on `aws_sns_topic`.  Each `aws_sns_topic_subscription` represents one email, phone number, SQS queue, etc... so while there may be a few defined in TF, if they are doing any kind of volume delivery the subscriptions will be created outside of IaC.
3. The current `aws_sns_topic` implementation assumes only https subscriptions, but we should add support for SMS and the four other charged subscriptions because the prices can be quite different.  We can add usage params so the number of subscribers of each type can be specified.
4. We should add support for FIFO topics too.

Going with that the cost components for Standard topics would be:
| Name                           | Unit               | Multiplier | Info                                                                                                          | Price Info                                                                                                                                               |
| ------------------------------ | ------------------ | ---------- | ------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
| API requests                   | 1M requests        | 1000000    | quantity from usage params monthly\_requests & request\_size                                                  | { product: "AmazonSNS" priceFamily: "API Request"} $0.50 per 1,000,000 Amazon SNS API Requests per month thereafter                                      |
| HTTP/HTTPS notifications       | 100k notifications | 100000     | quantity from usage: http\_subscribers \* monthly\_requests                                                   | { product: "AmazonSNS" priceFamily: "Message Delivery"} $0.06 per 100,000 Amazon SNS HTTP/HTTPS Notifications thereafter                                 |
| Email/Email-JSON notifications | 100k notifications | 100000     | quantity from usage: email\_subscribers \* monthly\_requests                                                  | { product: "AmazonSNS" priceFamily: "Message Delivery"} $2.00 per 100,000 Amazon SNS Email/Email-JSON Notifications thereafter                           |
| Kinesis Firehose notifications | 1M notifications   | 1000000    | quantity from usage: kinesis\_subscribers \* monthly\_requests                                                | { product: "AmazonSNS" priceFamily: "Message Delivery"} Message Delivery | $0.19 per 1,000,000 SNS to Kinesis Firehose messages in US East (N. Virginia) |
| Mobile Push notifications      | 1M notifications   | 1000000    | quantity from usage: mobile\_push\_subscribers \* monthly\_requests                                           | { product: "AmazonSNS" priceFamily: "Message Delivery"} Message Delivery | $0.50 per 1,000,000 Mobile Push Notifications per month thereafter            |
| MacOS notifications            | 1M notifications   | 1000000    | quantity from usage: macos\_subscribers \* monthly\_requests                                                  | { product: "AmazonSNS" priceFamily: "Message Delivery"} Message Delivery | $0.50 per 1,000,000 MacOS Notifications per month thereafter                  |
| SMS notifications              | 100 notifications  | 100        | quantity from usage: sms\_subscribers \* monthly\_requests; price overridable with usage sms\_price\_per\_100 | { product: "AmazonSNS" priceFamily: "Message Delivery"} Message Delivery | $0.75 per 100 Amazon SNS SMS Notifications thereafter                         |

Cost components for FIFO topics would be
| Name                      | Unit             | Multiplier | Info                                                                                                    | Price Info                                                                                                             |
| ------------------------- | ---------------- | ---------- | ------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| FIFO Publish API requests | 1M requests      | 1000000    | quantity from usage: monthly\_requests                                                                  | { product: "AmazonSNS" priceFamily: ""} $0.3 per 1,000,000 SNS FIFO Publish API Requests in US East (N. Virginia)      |
| FIFO Publish API payload  | GB               |            | quantity from usage: monthly\_requests \* request\_size\_kb                                             | { product: "AmazonSNS" priceFamily: ""} $0.017 per GB of SNS FIFO Publish Messages Payload in US East (N. Virginia)    |
| FIFO notifications        | 1M notifications | 1000000    | quantity from: monthly\_requests(usage) \* count of aws\_sns\_topic\_subscriptions                      | { product: "AmazonSNS" priceFamily: ""} $0.01 per 1,000,000 SNS FIFO Messages for Subscribers in US East (N. Virginia) |
| FIFO notification payload | GB               |            | quantity from: monthly\_requests(usage) \* count of aws\_sns\_topic\_subscriptions \* request\_size\_kb | { product: "AmazonSNS" priceFamily: ""} $0.001 per GB of SNS FIFO Messages for Subscribers in US East (N. Virginia)    |

The usage params would be
| Name                        | Type  | Default | Description                                                                           |
| --------------------------- | ----- | ------- | ------------------------------------------------------------------------------------- |
| monthly\_requests           | int   | 0       | \# Monthly requests to SNS.                                                           |
| request\_size\_kb           | float | 1       | \# Size of requests to SNS. Stanard billed in 64KB chunks, FIFO billed in 1KB chunks. |
| http\_subscriptions         | int   | 0       | \# Number of HTTP/HTTPS subscriptions                                                 |
| email\_subscriptions        | int   | 0       | \# Number of Email/Email-JSON subscriptions                                           |
| kinesis\_subscriptions      | int   | 0       | \# Number of Kinesis Firehose subscriptions                                           |
| mobile\_push\_subscriptions | int   | 0       | \# Number of Mobile Push subscriptions                                                |
| macos\_subscriptions        | int   | 0       | \# Number of MacOS subscriptions                                                      |
| sms\_subscriptions          | int   | 0       | \# Number of SMS subscriptions                                                        |
| sms\_notification\_price    | float | 0.0075  | \# Price per SMS notificatopm                                                         |

And the cost components and usage params for aws_sns_topic_subscription would be deprecated.

**Comment by hugorut** [comment]
@tim775 sounds good, agreed `aws_sns_topic` looks like the right place for these prices 👍

**Comment by hugorut** [comment]
This is fixed in the [0.9.22](https://github.com/infracost/infracost/releases/tag/v0.9.22) release.
